### PR TITLE
Add sax plugin for capacity and its subkeys

### PR DESCRIPTION
**Pull request by gileri**
This plugins check that :
* `capacity=*` and `capacity:*=*` values are positive integers
* `capacity:*=*` subvalues are inferior or equal to `capacity=*`

https://wiki.openstreetmap.org/wiki/Key:capacity

I picked 3500 as an item number, but I'm not sure how those should be allocated.

### CODE REVIEW

File: plugins/Capacity.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+    def node(self, data, tags):
+        try:
+            if tags["capacity"] == "":
+                # Ignore errors that should be reported by generic prefix analysers
+                return
+            total_capacity = int(tags["capacity"])
+        except KeyError:
+            # No capacity tag
+            return
```
Comment (frodrigo) [inline plugins/Capacity.py:46]: Could you just check for "capacity" in tags, rather than raisin an exception (to be the same as other plugins).

File: plugins/Capacity.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+        except ValueError:
+            return {
+                "class": 35001,
+                "subclass": 0,
+                "text": T_('Invalid "{0}" value: {1}', "capacity", tags["capacity"]),
+            }
```
Comment (frodrigo) [inline plugins/Capacity.py:52]: Maybe we can check this in Number plugin ?

**Comment by frodrigo** [comment]
Maybe add it in item 3091.

The sum of sub-capacity should lower then total capacity?

File: plugins/Capacity.py (hunk 1 above)
Comment (gileri) [reply plugins/Capacity.py]: Changed in be3f47bc.

**Comment by gileri** [comment]
> The sum of sub-capacity should lower then total capacity?

I don't think it should, usually the same spot/capacity allow for multiple user types, for example :

* bus, coach, hgv, trailer
* disabled and wheelchair

File: plugins/Capacity.py (hunk 2 above)
Comment (gileri) [reply plugins/Capacity.py:50]: I thought about adding valid integer and min/max validation, but it would introduce additional complexity to Number, which seems already not really [straightforward](https://github.com/osm-fr/osmose-backend/blob/8709c8e/plugins/Number.py#L60-L74). Also it currently allows having an height defined in kilograms ^^.

I tried defining a [declarative rule set](https://gitlab.com/-/snippets/2038252) for the Number plugin (or any plugin really), but while rules would be more explicit and precise, that would also make it tricky to handle edge-cases.

In conclusion I think it would be easier and clearer to handle each ~numerical "tag family" (speed, height, capacity etc.) in its own plugin/class to avoid convoluted case handling, even if that causes a bit of repetition.

What do you think ?

Comment (frodrigo) [reply plugins/Capacity.py:50]: MapCSS may be more appropriate for "declarative" rule set. JOSM already have a MapCSS for this.

File: plugins/Capacity.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+                or key == "capacity:"
+                or value == ""
+                or value in ("yes", "no")
+            ):
+                continue
+
+            try:
+                capacity_int = int(value)
+            except ValueError:
+                return {
```
Comment (frodrigo) [inline plugins/Capacity.py:72]: You may return an array, you do not have to return on first issue.

File: plugins/Capacity.py (hunk 2 above)
Comment (gileri) [reply plugins/Capacity.py:50]: JOSM does have some numeric validation in MapCSS, but not for `capacity`, `capacity:*`. Can you expand on what you think should be done ? Thank you :)

Comment (frodrigo) [reply plugins/Capacity.py:50]: The rule can (probably) also be written in MapCSS an shared with JOSM.

Comment (gileri) [reply plugins/Capacity.py:50]: I agree that declarative (and MapCSS should be preferred when possible. This particular error could indeed be handled by a MapCSS rule. However :

* The plugin must do this check anyway to compare the value to sub-capacities later. Why not raise the error ?
* [MapCSS don't support wildcard/regex for keys](https://wiki.openstreetmap.org/wiki/MapCSS/0.2#Tests), so the same check for subcapacities must be implemented in this plugin anyway.
* That would split error handling on this tag between two plugins.

Do you think I should make a separate MapCSS and Plugin containing this check only ? I don't mind as it seems easy enough thanks to [mapcss2osmose.py](https://github.com/osm-fr/osmose-backend/blob/master/mapcss/mapcss2osmose.py).

Comment (frodrigo) [reply plugins/Capacity.py:50]: So ok, we keep think like this.

File: plugins/Capacity.py (hunk 3 above)
Comment (gileri) [reply plugins/Capacity.py]: Thank you for the suggestion, I made it so in 3ddfae591179175b88d8e3156417625157bafd76.

**Comment by gileri** [comment]
> Maybe add it in item 3091.

That look fitting indeed, I changed the error definitions in 9bb8e343385659a278a121d3a3c8470dd1090f3d.

File: plugins/Capacity.py [hunk 4]
```diff
@@ -26,79 +26,90 @@ class Capacity(Plugin):
...
-            detail=T_("""A capacity tag value is incorrect."""),
+            detail=T_("""Capacity tags should be positive integers."""),
+            resource="https://wiki.openstreetmap.org/wiki/Key:capacity",
+        )
+        self.errors[30913] = self.def_class(
+            item=3091,
+            level=2,
+            tags=["tag"],
+            title=T_("Specific capacity is greater than total capacity"),
+            detail=T_("""A capacity:* value is greater than the total capacity."""),
```
Comment (frodrigo) [inline plugins/Capacity.py:42]:
```suggestion
            detail=T_("""A capacity:* value should be lower than the total capacity."""),
```

File: plugins/Capacity.py [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+            total_capacity = None
+
+        for key, value in tags.items():
+            if (
+                not key.startswith("capacity:")
+                # Ignore errors that should be reported by generic prefix analysers
+                or key == "capacity:"
+                or value == ""
+
+                or value in ("yes", "no")
```
Comment (frodrigo) [inline plugins/Capacity.py:75]:
```suggestion
                or value in ("", "yes", "no")
```

File: plugins/Capacity.py (hunk 4 above)
Comment (gileri) [reply plugins/Capacity.py:42]: Changed in b939fc85dc648b50702dc4d8c2b1007fb39c3310

File: plugins/Capacity.py (hunk 5 above)
Comment (gileri) [reply plugins/Capacity.py:75]: I split those two checks as `""` is an invalid value that should be checked in other Analysers, while `yes|no` are valid values. I believe the performance penalty is really low for the clarity added by [the comment L71](https://github.com/osm-fr/osmose-backend/pull/1023/files#diff-d376487ccb6816192f500423b8279401ea86f138eb7be2bbb05c26567be043e9R71-R73).

Comment (frodrigo) [reply plugins/Capacity.py:75]: Ok. So just remove the empty line, please.

Comment (gileri) [reply plugins/Capacity.py:75]: I reordered the conditions to remove the empty line while still keeping the comment relevant, I hope that's okay.

File: plugins/Capacity.py [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+            })
+            total_capacity = None
+
+        for key, value in tags.items():
+            if (
+                not key.startswith("capacity:")
+                # Ignore errors that should be reported by generic prefix analysers
+                or key == "capacity:"
+                or value == ""
+
```
Comment (frodrigo) [inline plugins/Capacity.py:74]:
```suggestion
```

**Comment by frodrigo** [comment]
Merged, thank you.

**Comment by gileri** [comment]
Thank you @frodrigo for the review and merge !

**Comment by frodrigo** [comment]
Deployed.

**Comment by Famlam** [comment]
Added a minor fix for man_made=* keys, where it's commonly used for volumes. (And volumes of course can be decimal and/or with units) #1398
